lib/mariaex/messages.ex

```diff
@@ -264,13 +233,103 @@ defmodule Mariaex.Messages do
     << acc :: binary, 0 :: size(padding), reversebits :: bits >>
   end
 
+  # Decoding
+
+  def decode(<< len :: size(24)-little-integer, seqnum :: size(8)-integer, body :: size(len)-binary, rest :: binary>>, state),
+    do: {packet(size: len, seqnum: seqnum, msg: decode_msg(body, state), body: body), rest}
+  def decode(rest, _state), 
+    do: {nil, rest}
+ 
+  defp decode_msg(body, :handshake),                                               do: __decode__(:handshake, body)
+  defp decode_msg(<< 0 :: 8, _ :: binary >> = body, :bin_rows),                    do: __decode__(:bin_row, body)
+  defp decode_msg(<< 0 :: 8, _ :: binary >> = body, :prepare_send),                do: __decode__(:stmt_prepare_ok, body)
+  defp decode_msg(<< 0 :: 8, _ :: binary >> = body, _),                            do: __decode__(:ok_resp, body)
+  defp decode_msg(<< 254 :: 8, _ :: binary >> = body, _) when byte_size(body) < 9, do: __decode__(:eof_resp, body)
+  defp decode_msg(<< 255 :: 8, _ :: binary >> = body, _),                          do: __decode__(:error_resp, body)
+  defp decode_msg(body, :column_count),                                            do: __decode__(:column_count, body)
+  defp decode_msg(body, :column_definitions),                                      do: __decode__(:column_definition_41, body)
+  defp decode_msg(body, :rows),                                                    do: __decode__(:row, body)
+
+  defp decode_string(data),    do: data
+  defp decode_float(data),     do: Float.parse(data) |> elem(0)
+  defp decode_integer(data),   do: Integer.parse(data) |> elem(0)
+  defp decode_bit(<<bit>>),    do: bit
+  defp decode_null(_),         do: nil
+
+  defp decode_date(data) do 
+    [year, month, day] = '~d-~d-~d'
```

Comment:
I think String.split(data, "-") |> Enum.map(&String.to_integer/1) will be cleaner and faster!